tests/test_directives.py

```diff
@@ -0,0 +1,270 @@
+# Copyright (c) 2017 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+import pytest
+import sys
+
+from flow.directives import Directives, _DirectivesItem
+from flow.directives import NP, NRANKS, NGPU, EXECUTABLE, OMP_NUM_THREADS
+from flow.directives import WALLTIME, MEMORY, PROCESS_FRACTION
+from flow.directives import no_aggregation
+from flow import FlowProject
+from tempfile import TemporaryDirectory
+
+
+@pytest.fixture()
+def available_directives_list():
+    return [NP, NRANKS, NGPU, OMP_NUM_THREADS,
+            EXECUTABLE, WALLTIME, MEMORY,
+            PROCESS_FRACTION]
+
+
+@pytest.fixture()
+def product_directive():
+    def val(v):
+        if v < 10:
+            raise ValueError("Price cannot be less than 10 units")
+        return v
+
+    def finalize(value, dict):
+        discount = dict.get('discount', 0)
+        free = dict.get('free', False)
+        value = value - discount
+        if value < 0 or free:
+            return 0
+        return value
+
+    product = _DirectivesItem(name='product', validation=val,
+                              default=10, serial=no_aggregation,
+                              parallel=no_aggregation, finalize=finalize)
+    return product
+
+
+@pytest.fixture()
+def non_default_directive_values():
+    def _values(values=0):
+        if values == 1:
+            return {'np': 4, 'ngpu': 1, 'nranks': 0,
+                    'omp_num_threads': 10, 'executable': 'PathFinder',
+                    'walltime': 20., 'memory': 16, 'processor_fraction': 0.5}
+        return {'np': 4, 'ngpu': 10, 'nranks': 5,
+                'omp_num_threads': 20, 'executable': 'Non Default Path',
+                'walltime': 64., 'memory': 32, 'processor_fraction': 0.5}
+    return _values
+
+
+class TestItems:
+    """
+    Tests for _DirectivesItem class
+    """
+    def test_default(self):
+        assert NP.default == 1
+        assert NGPU.default == 0
+        assert NRANKS.default == 0
+        assert OMP_NUM_THREADS.default == 0
+        assert EXECUTABLE.default == sys.executable
+        assert WALLTIME.default == 12.0
+        assert MEMORY.default == 4
+        assert PROCESS_FRACTION.default == 1.
+
+    def test_invalid_values(self, available_directives_list):
```

Comment:
This is a good test, I would potentially add more invalid_values (e.g. `'np': [0, -1, 'foo', {}]`). To make it easier for you in this case, I would just test that the invalid values raise and not worry about the exception type.

Reply:
That'd be great, I'll do it.